src/data/list/rotate.lean

```diff
@@ -102,7 +111,233 @@ lemma prod_rotate_eq_one_of_prod_eq_one [group α] : ∀ {l : list α} (hl : l.p
 | (a::l) hl n :=
 have n % list.length (a :: l) ≤ list.length (a :: l), from le_of_lt (nat.mod_lt _ dec_trivial),
 by rw ← list.take_append_drop (n % list.length (a :: l)) (a :: l) at hl;
-  rw [← rotate_mod, rotate_eq_take_append_drop this, list.prod_append, mul_eq_one_iff_inv_eq,
+  rw [← rotate_mod, rotate_eq_drop_append_take this, list.prod_append, mul_eq_one_iff_inv_eq,
     ← one_mul (list.prod _)⁻¹, ← hl, list.prod_append, mul_assoc, mul_inv_self, mul_one]
 
+lemma rotate_perm (l : list α) (n : ℕ) : l.rotate n ~ l :=
+begin
+  rw rotate_eq_rotate',
+  induction n with n hn generalizing l,
+  { simp },
+  { cases l with hd tl,
+    { simp },
+    { rw rotate'_cons_succ,
+      exact (hn _).trans (perm_append_singleton _ _) } }
+end
+
+@[simp] lemma nodup_rotate {l : list α} {n : ℕ} : nodup (l.rotate n) ↔ nodup l :=
+(rotate_perm l n).nodup_iff
+
+@[simp] lemma rotate_eq_nil_iff {l : list α} {n : ℕ} : l.rotate n = [] ↔ l = [] :=
+begin
+  induction n with n hn generalizing l,
+  { simp },
+  { cases l with hd tl,
+    { simp },
+    { simp [rotate_cons_succ, hn] } }
+end
+
+@[simp] lemma rotate_singleton (x : α) (n : ℕ) :
+  [x].rotate n = [x] :=
+begin
+  induction n with n hn,
+  { simp },
+  { rwa [rotate_cons_succ] }
+end
+
+lemma zip_with_rotate_distrib {α β γ : Type*} (f : α → β → γ) (l : list α) (l' : list β) (n : ℕ)
+  (h : l.length = l'.length) :
+  (zip_with f l l').rotate n = zip_with f (l.rotate n) (l'.rotate n) :=
+begin
+  rw [rotate_eq_drop_append_take_mod, rotate_eq_drop_append_take_mod,
+      rotate_eq_drop_append_take_mod, h, zip_with_append, ←zip_with_distrib_drop,
+      ←zip_with_distrib_take, list.length_zip_with, h, min_self],
+  rw [length_drop, length_drop, h]
+end
+
+local attribute [simp] rotate_cons_succ
+
+@[simp] lemma zip_with_rotate_one {β : Type*} (f : α → α → β) (x y : α) (l : list α) :
+  zip_with f (x :: y :: l) ((x :: y :: l).rotate 1) =
+  f x y :: zip_with f (y :: l) (l ++ [x]) :=
+by simp
+
+lemma nth_le_rotate_one (l : list α) (k : ℕ) (hk : k < (l.rotate 1).length) :
+  (l.rotate 1).nth_le k hk = l.nth_le ((k + 1) % l.length)
+    (by { cases l, { simpa using hk }, { refine nat.mod_lt _ _, simp }}) :=
```

Comment:
From what I remember, putting the proof inline like this prevents rw and/or simp from using this lemma when passed backwards. I think something like this will fix it:

```suggestion
lemma nth_le_rotate_one (l : list α) (k : ℕ) (hk : k < (l.rotate 1).length)
  (hk_mod : sorry := by { cases l, { simpa using hk }, { refine nat.mod_lt _ _, simp }}):
  (l.rotate 1).nth_le k hk = l.nth_le ((k + 1) % l.length) hk_mod :=
```

Reply:
Is this the case for any tactic proof? In my brief tests with this lemma, I've been able to rewrite backwards even if the proof witness is via tactic mode.
I've made a change to make the witness be in term mode, which I think does end up working with rewrites. Do you have an example of how this lemma cannot be used in a reverse rw or simp?

Reply:
I think `simp` and `rw` know about proof irrelevance. Here's a rather artificial test case:

```lean
inductive foo : Type | one : foo

-- Definitely not going to be equal to something else unless you have proof irrelevance:
axiom hfoo : ∀ (x y : foo), x = y

def bar {x y : foo} (h : x = y) : foo := x

lemma bar_refl {x : foo} : bar (refl x) = bar (hfoo x x) := rfl
lemma bar_hfoo {x y : foo} : bar (hfoo x y) = x := rfl

example {x : foo} (hxy : x = x) : bar hxy = x := by rw [← bar_refl, bar_hfoo]
example {x : foo} (hxy : x = x) : bar hxy = x := by simp only [← bar_refl, bar_hfoo]
```

Reply:
Tactic mode is not the issue here, but I was under the impression that this did matter, and have made changes [like this one](https://github.com/leanprover-community/mathlib/pull/6479/files#diff-3305c14b38e4b6d201198ab5c73f7dd140ee05cefa0d47c5af0be4d85137e513R242) which I thought fixed an issue here.

Perhaps @gebner or @digama0 can remind me what the issue I'm thinking of is?

Reply:
I've seen this pattern sometimes (generalizing proof terms into arguments), but I don't think it is necessary.  (At least I'll continue to believe so until I see a concrete example where it causes problems.)